File: Training/Checkpointing/pi_requeue.py
#!/usr/bin/env python3

"""
pi_requeue.py

Monte Carlo estimation of pi designed for Slurm requeue partitions.

Features
--------
- periodic application-level checkpoints
- automatic checkpoint discovery and restart
- exact RNG-state restoration
- atomic checkpoint writes
- SIGUSR1 and SIGTERM handling
- awareness of Slurm job/restart information

Unlike pi_checkpoint.py and pi_signal.py, no --resume option is needed.
If a checkpoint exists, the program automatically resumes from it.

Use --fresh to deliberately start over.
"""

import argparse
import math
import os
import pickle
import random
import signal
import sys
import time
from pathlib import Path


CHECKPOINT_VERSION = 1

stop_requested = False
received_signal = None


# ----------------------------------------------------------------------
# Signal handling
# ----------------------------------------------------------------------

def signal_handler(signum, frame):
    """
    Keep the signal handler minimal.

    The handler only records that termination was requested.
    Checkpoint I/O is performed later from the normal program flow.
    """
    global stop_requested
    global received_signal

    stop_requested = True
    received_signal = signum


# ----------------------------------------------------------------------
# CLI
# ----------------------------------------------------------------------

def parse_args():

    parser = argparse.ArgumentParser(
        description=(
            "Monte Carlo pi calculation with automatic "
            "checkpoint/restart for Slurm requeue jobs."
        )
    )

    parser.add_argument(
        "-n",
        "--darts",
        type=int,
        default=100_000_000,
        help="Total number of darts",
    )

    parser.add_argument(
        "--seed",
        type=int,
        default=42,
        help="Random-number generator seed",
    )

    parser.add_argument(
        "--report-every",
        type=int,
        default=1_000_000,
        help="Print progress every N darts",
    )

    parser.add_argument(
        "--checkpoint-every",
        type=int,
        default=5_000_000,
        help="Write a periodic checkpoint every N darts",
    )

    parser.add_argument(
        "--checkpoint-file",
        type=Path,
        default=Path("pi_checkpoint.pkl"),
        help="Checkpoint filename",
    )

    parser.add_argument(
        "--sleep",
        type=float,
        default=0.0,
        help="Optional delay after progress reports",
    )

    parser.add_argument(
        "--fresh",
        action="store_true",
        help="Ignore and remove any existing checkpoint",
    )

    return parser.parse_args()


# ----------------------------------------------------------------------
# Checkpoint functions
# ----------------------------------------------------------------------

def save_checkpoint(
    checkpoint_file,
    completed_darts,
    inside_circle,
    rng,
    n_darts,
    seed,
    elapsed,
):

    checkpoint = {
        "version": CHECKPOINT_VERSION,
        "completed_darts": completed_darts,
        "inside_circle": inside_circle,
        "rng_state": rng.getstate(),
        "n_darts": n_darts,
        "seed": seed,
        "elapsed": elapsed,
    }

    checkpoint_file.parent.mkdir(
        parents=True,
        exist_ok=True,
    )

    tmp_file = checkpoint_file.with_name(
        checkpoint_file.name + ".tmp"
    )

    with open(tmp_file, "wb") as f:

        pickle.dump(
            checkpoint,
            f,
            protocol=pickle.HIGHEST_PROTOCOL,
        )

        f.flush()
        os.fsync(f.fileno())

    os.replace(
        tmp_file,
        checkpoint_file,
    )


def load_checkpoint(
    checkpoint_file,
    expected_darts,
    expected_seed,
):

    with open(checkpoint_file, "rb") as f:
        checkpoint = pickle.load(f)

    if checkpoint.get("version") != CHECKPOINT_VERSION:
        raise RuntimeError(
            "Checkpoint version is incompatible."
        )

    if checkpoint["n_darts"] != expected_darts:
        raise RuntimeError(
            "Checkpoint belongs to a different calculation:\n"
            f"  checkpoint darts : {checkpoint['n_darts']:,}\n"
            f"  requested darts  : {expected_darts:,}"
        )

    if checkpoint["seed"] != expected_seed:
        raise RuntimeError(
            "Checkpoint uses a different RNG seed:\n"
            f"  checkpoint seed : {checkpoint['seed']}\n"
            f"  requested seed  : {expected_seed}"
        )

    return checkpoint


# ----------------------------------------------------------------------
# Calculation
# ----------------------------------------------------------------------

def estimate_pi(args):

    rng = random.Random(args.seed)

    completed_darts = 0
    inside_circle = 0
    previous_elapsed = 0.0

    #
    # Automatic restart
    #
    if args.checkpoint_file.exists():

        checkpoint = load_checkpoint(
            checkpoint_file=args.checkpoint_file,
            expected_darts=args.darts,
            expected_seed=args.seed,
        )

        completed_darts = checkpoint["completed_darts"]
        inside_circle = checkpoint["inside_circle"]
        previous_elapsed = checkpoint["elapsed"]

        rng.setstate(
            checkpoint["rng_state"]
        )

        print()
        print("=" * 72)
        print("CHECKPOINT FOUND - AUTOMATIC RESTART")
        print("=" * 72)
        print(
            f"Completed darts  : "
            f"{completed_darts:,}"
        )
        print(
            f"Inside circle    : "
            f"{inside_circle:,}"
        )
        print(
            f"Previous runtime : "
            f"{previous_elapsed:.2f} s"
        )
        print("=" * 72)
        print()

        #
        # It is possible that a completed calculation
        # is being submitted again.
        #
        if completed_darts >= args.darts:

            pi_estimate = (
                4.0
                * inside_circle
                / completed_darts
            )

            return (
                pi_estimate,
                inside_circle,
                previous_elapsed,
                True,
            )

    else:

        print()
        print("No checkpoint found.")
        print("Starting a new calculation.")
        print()

    start_time = time.perf_counter()

    #
    # Main Monte Carlo loop
    #
    for i in range(
        completed_darts + 1,
        args.darts + 1,
    ):

        x = rng.uniform(-1.0, 1.0)
        y = rng.uniform(-1.0, 1.0)

        if x * x + y * y <= 1.0:
            inside_circle += 1

        #
        # Signal-triggered checkpoint
        #
        if stop_requested:

            elapsed = (
                previous_elapsed
                + time.perf_counter()
                - start_time
            )

            signal_name = signal.Signals(
                received_signal
            ).name

            print()
            print("=" * 72)
            print("SIGNAL RECEIVED")
            print("=" * 72)
            print(
                f"Signal           : {signal_name}"
            )
            print(
                f"Completed darts  : {i:,}"
            )
            print("Saving checkpoint ...")

            save_checkpoint(
                checkpoint_file=args.checkpoint_file,
                completed_darts=i,
                inside_circle=inside_circle,
                rng=rng,
                n_darts=args.darts,
                seed=args.seed,
                elapsed=elapsed,
            )

            print(
                f"Checkpoint saved : "
                f"{args.checkpoint_file}"
            )

            print("=" * 72)

            return None

        #
        # Progress report
        #
        if (
            args.report_every > 0
            and (
                i % args.report_every == 0
                or i == args.darts
            )
        ):

            pi_estimate = (
                4.0 * inside_circle / i
            )

            elapsed = (
                previous_elapsed
                + time.perf_counter()
                - start_time
            )

            print(
                f"darts = {i:>12,d} / "
                f"{args.darts:,}   "
                f"pi = {pi_estimate:.8f}   "
                f"error = "
                f"{abs(pi_estimate - math.pi):.3e}   "
                f"elapsed = {elapsed:.2f} s",
                flush=True,
            )

            if args.sleep > 0:
                time.sleep(args.sleep)

        #
        # Periodic checkpoint
        #
        if (
            args.checkpoint_every > 0
            and i % args.checkpoint_every == 0
        ):

            elapsed = (
                previous_elapsed
                + time.perf_counter()
                - start_time
            )

            save_checkpoint(
                checkpoint_file=args.checkpoint_file,
                completed_darts=i,
                inside_circle=inside_circle,
                rng=rng,
                n_darts=args.darts,
                seed=args.seed,
                elapsed=elapsed,
            )

            print(
                f"  --> periodic checkpoint "
                f"saved at dart {i:,}",
                flush=True,
            )

    #
    # Normal completion
    #
    total_elapsed = (
        previous_elapsed
        + time.perf_counter()
        - start_time
    )

    pi_estimate = (
        4.0
        * inside_circle
        / args.darts
    )

    #
    # Save the completed state.
    #
    save_checkpoint(
        checkpoint_file=args.checkpoint_file,
        completed_darts=args.darts,
        inside_circle=inside_circle,
        rng=rng,
        n_darts=args.darts,
        seed=args.seed,
        elapsed=total_elapsed,
    )

    return (
        pi_estimate,
        inside_circle,
        total_elapsed,
        False,
    )


# ----------------------------------------------------------------------
# Main
# ----------------------------------------------------------------------

def main():

    args = parse_args()

    if args.darts <= 0:
        raise ValueError(
            "--darts must be greater than zero"
        )

    if args.report_every < 0:
        raise ValueError(
            "--report-every cannot be negative"
        )

    if args.checkpoint_every < 0:
        raise ValueError(
            "--checkpoint-every cannot be negative"
        )

    if args.sleep < 0:
        raise ValueError(
            "--sleep cannot be negative"
        )

    #
    # Remove previous state only when explicitly requested.
    #
    if args.fresh and args.checkpoint_file.exists():

        print(
            f"Removing existing checkpoint: "
            f"{args.checkpoint_file}"
        )

        args.checkpoint_file.unlink()

    #
    # Register signals.
    #
    signal.signal(
        signal.SIGUSR1,
        signal_handler,
    )

    signal.signal(
        signal.SIGTERM,
        signal_handler,
    )

    #
    # Slurm information
    #
    job_id = os.environ.get(
        "SLURM_JOB_ID",
        "not-running-under-slurm",
    )

    restart_count = os.environ.get(
        "SLURM_RESTART_COUNT",
        "0",
    )

    hostname = os.environ.get(
        "HOSTNAME",
        "unknown",
    )

    print("=" * 72)
    print(
        "Monte Carlo Pi - "
        "Automatic Slurm Checkpoint/Restart"
    )
    print("=" * 72)

    print(f"Slurm job ID        : {job_id}")
    print(f"Slurm restart count : {restart_count}")
    print(f"Hostname            : {hostname}")

    print("-" * 72)

    print(
        f"Number of darts     : "
        f"{args.darts:,}"
    )

    print(
        f"Random seed         : "
        f"{args.seed}"
    )

    print(
        f"Checkpoint interval : "
        f"{args.checkpoint_every:,}"
    )

    print(
        f"Checkpoint file     : "
        f"{args.checkpoint_file}"
    )

    print("=" * 72)

    result = estimate_pi(args)

    #
    # Signal-triggered exit.
    #
    if result is None:

        print()
        print(
            "Application checkpointed "
            "and is terminating."
        )

        sys.exit(0)

    (
        pi_estimate,
        inside_circle,
        elapsed,
        already_complete,
    ) = result

    print()
    print("=" * 72)

    if already_complete:
        print(
            "Calculation was already complete."
        )
    else:
        print("CALCULATION COMPLETE")

    print("=" * 72)

    print(
        f"Darts thrown       : "
        f"{args.darts:,}"
    )

    print(
        f"Inside circle      : "
        f"{inside_circle:,}"
    )

    print(
        f"Estimated pi       : "
        f"{pi_estimate:.10f}"
    )

    print(
        f"Python math.pi     : "
        f"{math.pi:.10f}"
    )

    print(
        f"Absolute error     : "
        f"{abs(pi_estimate - math.pi):.6e}"
    )

    print(
        f"Total elapsed time : "
        f"{elapsed:.2f} s"
    )

    print("=" * 72)


if __name__ == "__main__":
    main()

File: Training/Checkpointing/Exercise2/pi_requeue.c
/*
 * pi_requeue.c
 *
 * Monte Carlo pi estimation designed for Slurm requeue jobs.
 *
 * Features:
 *   - periodic checkpointing
 *   - automatic checkpoint discovery
 *   - automatic restart
 *   - RNG-state restoration
 *   - SIGUSR1/SIGTERM support
 *   - SLURM_JOB_ID and SLURM_RESTART_COUNT reporting
 *
 * Intended for CANNON serial_requeue.
 */

#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <math.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>


#define CHECKPOINT_VERSION 1

static const double PI_TRUE =
    3.141592653589793238462643383279502884;


static volatile sig_atomic_t stop_requested = 0;
static volatile sig_atomic_t received_signal = 0;


typedef struct {
    char magic[8];
    uint32_t version;

    uint64_t completed_darts;
    uint64_t inside_circle;
    uint64_t rng_state;

    uint64_t total_darts;
    uint64_t seed;

    double elapsed_seconds;
} checkpoint_t;


typedef struct {
    uint64_t darts;
    uint64_t seed;
    uint64_t report_every;
    uint64_t checkpoint_every;

    double sleep_seconds;

    const char *checkpoint_file;

    int fresh;
} options_t;


/* ------------------------------------------------------------------ */
/* Signal handling                                                    */
/* ------------------------------------------------------------------ */

static void signal_handler(int signum)
{
    stop_requested = 1;
    received_signal = signum;
}


static void install_signal_handlers(void)
{
    struct sigaction action;

    memset(&action, 0, sizeof(action));

    action.sa_handler = signal_handler;
    action.sa_flags = 0;

    if (sigemptyset(&action.sa_mask) != 0) {
        perror("sigemptyset");
        exit(EXIT_FAILURE);
    }

    if (sigaction(SIGUSR1, &action, NULL) != 0) {
        perror("sigaction SIGUSR1");
        exit(EXIT_FAILURE);
    }

    if (sigaction(SIGTERM, &action, NULL) != 0) {
        perror("sigaction SIGTERM");
        exit(EXIT_FAILURE);
    }
}


/* ------------------------------------------------------------------ */
/* RNG                                                                */
/* ------------------------------------------------------------------ */

static uint64_t rng_next(uint64_t *state)
{
    uint64_t x = *state;

    x ^= x >> 12;
    x ^= x << 25;
    x ^= x >> 27;

    *state = x;

    return x * UINT64_C(2685821657736338717);
}


static double rng_uniform(uint64_t *state)
{
    return (rng_next(state) >> 11) *
           (1.0 / 9007199254740992.0);
}


/* ------------------------------------------------------------------ */
/* Timing                                                             */
/* ------------------------------------------------------------------ */

static double wall_time(void)
{
    struct timespec ts;

    if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
        perror("clock_gettime");
        exit(EXIT_FAILURE);
    }

    return (double) ts.tv_sec +
           (double) ts.tv_nsec * 1.0e-9;
}


static void sleep_seconds(double seconds)
{
    if (seconds <= 0.0) {
        return;
    }

    struct timespec req;

    req.tv_sec = (time_t) seconds;
    req.tv_nsec =
        (long) ((seconds - (double) req.tv_sec) * 1.0e9);

    while (nanosleep(&req, &req) != 0) {

        if (errno == EINTR && stop_requested) {
            return;
        }

        if (errno != EINTR) {
            perror("nanosleep");
            exit(EXIT_FAILURE);
        }
    }
}


/* ------------------------------------------------------------------ */
/* Checkpoint I/O                                                     */
/* ------------------------------------------------------------------ */

static void initialize_checkpoint(checkpoint_t *checkpoint)
{
    memset(checkpoint, 0, sizeof(*checkpoint));

    memcpy(checkpoint->magic, "PICHECK", 7);

    checkpoint->version = CHECKPOINT_VERSION;
}


static int checkpoint_is_valid(const checkpoint_t *checkpoint)
{
    return (
        memcmp(checkpoint->magic, "PICHECK", 7) == 0 &&
        checkpoint->version == CHECKPOINT_VERSION
    );
}


static int file_exists(const char *filename)
{
    return access(filename, F_OK) == 0;
}


static void save_checkpoint(
    const char *filename,
    const checkpoint_t *checkpoint
)
{
    char tmp_filename[4096];

    int n = snprintf(
        tmp_filename,
        sizeof(tmp_filename),
        "%s.tmp",
        filename
    );

    if (n < 0 || (size_t) n >= sizeof(tmp_filename)) {
        fprintf(stderr, "Checkpoint filename too long.\n");
        exit(EXIT_FAILURE);
    }

    FILE *fp = fopen(tmp_filename, "wb");

    if (fp == NULL) {
        perror("fopen checkpoint");
        exit(EXIT_FAILURE);
    }

    if (fwrite(
            checkpoint,
            sizeof(*checkpoint),
            1,
            fp
        ) != 1) {

        perror("fwrite checkpoint");
        fclose(fp);
        exit(EXIT_FAILURE);
    }

    if (fflush(fp) != 0) {
        perror("fflush checkpoint");
        fclose(fp);
        exit(EXIT_FAILURE);
    }

    if (fsync(fileno(fp)) != 0) {
        perror("fsync checkpoint");
        fclose(fp);
        exit(EXIT_FAILURE);
    }

    if (fclose(fp) != 0) {
        perror("fclose checkpoint");
        exit(EXIT_FAILURE);
    }

    if (rename(tmp_filename, filename) != 0) {
        perror("rename checkpoint");
        exit(EXIT_FAILURE);
    }
}


static checkpoint_t load_checkpoint(const char *filename)
{
    checkpoint_t checkpoint;

    FILE *fp = fopen(filename, "rb");

    if (fp == NULL) {
        perror("fopen checkpoint");
        exit(EXIT_FAILURE);
    }

    if (fread(
            &checkpoint,
            sizeof(checkpoint),
            1,
            fp
        ) != 1) {

        fprintf(stderr, "Failed to read checkpoint.\n");
        fclose(fp);
        exit(EXIT_FAILURE);
    }

    fclose(fp);

    if (!checkpoint_is_valid(&checkpoint)) {
        fprintf(stderr, "Invalid checkpoint.\n");
        exit(EXIT_FAILURE);
    }

    return checkpoint;
}


/* ------------------------------------------------------------------ */
/* CLI                                                                */
/* ------------------------------------------------------------------ */

static uint64_t parse_uint64(
    const char *text,
    const char *name
)
{
    char *end = NULL;

    errno = 0;

    unsigned long long value =
        strtoull(text, &end, 10);

    if (errno != 0 || end == text || *end != '\0') {
        fprintf(stderr, "Invalid %s: %s\n", name, text);
        exit(EXIT_FAILURE);
    }

    return (uint64_t) value;
}


static double parse_double(
    const char *text,
    const char *name
)
{
    char *end = NULL;

    errno = 0;

    double value = strtod(text, &end);

    if (errno != 0 || end == text || *end != '\0') {
        fprintf(stderr, "Invalid %s: %s\n", name, text);
        exit(EXIT_FAILURE);
    }

    return value;
}


static options_t parse_args(int argc, char **argv)
{
    options_t options = {
        .darts = UINT64_C(100000000),
        .seed = UINT64_C(42),
        .report_every = UINT64_C(1000000),
        .checkpoint_every = UINT64_C(5000000),
        .sleep_seconds = 0.0,
        .checkpoint_file = "pi_requeue.bin",
        .fresh = 0
    };

    enum {
        OPT_SEED = 1000,
        OPT_REPORT,
        OPT_CKPT_EVERY,
        OPT_CKPT_FILE,
        OPT_SLEEP,
        OPT_FRESH
    };

    static const struct option long_options[] = {
        {"darts", required_argument, NULL, 'n'},
        {"seed", required_argument, NULL, OPT_SEED},
        {"report-every", required_argument, NULL, OPT_REPORT},
        {"checkpoint-every", required_argument, NULL, OPT_CKPT_EVERY},
        {"checkpoint-file", required_argument, NULL, OPT_CKPT_FILE},
        {"sleep", required_argument, NULL, OPT_SLEEP},
        {"fresh", no_argument, NULL, OPT_FRESH},
        {"help", no_argument, NULL, 'h'},
        {NULL, 0, NULL, 0}
    };

    int option;

    while ((option = getopt_long(
                argc,
                argv,
                "n:h",
                long_options,
                NULL
            )) != -1) {

        switch (option) {

            case 'n':
                options.darts =
                    parse_uint64(optarg, "--darts");
                break;

            case OPT_SEED:
                options.seed =
                    parse_uint64(optarg, "--seed");
                break;

            case OPT_REPORT:
                options.report_every =
                    parse_uint64(optarg, "--report-every");
                break;

            case OPT_CKPT_EVERY:
                options.checkpoint_every =
                    parse_uint64(optarg, "--checkpoint-every");
                break;

            case OPT_CKPT_FILE:
                options.checkpoint_file = optarg;
                break;

            case OPT_SLEEP:
                options.sleep_seconds =
                    parse_double(optarg, "--sleep");
                break;

            case OPT_FRESH:
                options.fresh = 1;
                break;

            case 'h':
                printf(
                    "Usage: %s [OPTIONS]\n"
                    "  --darts N\n"
                    "  --seed N\n"
                    "  --report-every N\n"
                    "  --checkpoint-every N\n"
                    "  --checkpoint-file FILE\n"
                    "  --sleep SEC\n"
                    "  --fresh\n",
                    argv[0]
                );
                exit(EXIT_SUCCESS);

            default:
                exit(EXIT_FAILURE);
        }
    }

    return options;
}


/* ------------------------------------------------------------------ */
/* Main                                                               */
/* ------------------------------------------------------------------ */

int main(int argc, char **argv)
{
    options_t options = parse_args(argc, argv);

    install_signal_handlers();

    /*
     * --fresh is useful interactively, but it should NOT normally
     * appear inside a requeue batch script.
     */
    if (options.fresh) {

        if (unlink(options.checkpoint_file) != 0 &&
            errno != ENOENT) {

            perror("unlink checkpoint");
            return EXIT_FAILURE;
        }
    }

    const char *job_id = getenv("SLURM_JOB_ID");
    const char *restart_count =
        getenv("SLURM_RESTART_COUNT");

    if (job_id == NULL) {
        job_id = "not-running-under-slurm";
    }

    if (restart_count == NULL) {
        restart_count = "0";
    }

    printf("========================================================================\n");
    printf("Monte Carlo Pi - C Slurm Requeue Example\n");
    printf("========================================================================\n");
    printf("Slurm job ID        : %s\n", job_id);
    printf("Slurm restart count : %s\n", restart_count);
    printf("Checkpoint file     : %s\n", options.checkpoint_file);
    printf("========================================================================\n");

    uint64_t completed_darts = 0;
    uint64_t inside_circle = 0;
    uint64_t rng_state = options.seed;

    double previous_elapsed = 0.0;

    /*
     * Automatic restart.
     */
    if (file_exists(options.checkpoint_file)) {

        checkpoint_t checkpoint =
            load_checkpoint(options.checkpoint_file);

        if (
            checkpoint.total_darts != options.darts ||
            checkpoint.seed != options.seed
        ) {

            fprintf(
                stderr,
                "Checkpoint configuration does not match this run.\n"
            );

            return EXIT_FAILURE;
        }

        completed_darts = checkpoint.completed_darts;
        inside_circle = checkpoint.inside_circle;
        rng_state = checkpoint.rng_state;
        previous_elapsed = checkpoint.elapsed_seconds;

        printf("\n");
        printf("CHECKPOINT FOUND - AUTOMATIC RESTART\n");
        printf("Completed darts  : %" PRIu64 "\n", completed_darts);
        printf("Inside circle    : %" PRIu64 "\n", inside_circle);
        printf("Previous runtime : %.2f s\n\n", previous_elapsed);

        if (completed_darts >= options.darts) {

            double pi_estimate =
                4.0 * (double) inside_circle /
                (double) completed_darts;

            printf("Calculation already complete.\n");
            printf("Estimated pi : %.10f\n", pi_estimate);

            return EXIT_SUCCESS;
        }

    } else {

        printf("\nNo checkpoint found.\n");
        printf("Starting a new calculation.\n\n");
    }

    const double start_time = wall_time();

    for (
        uint64_t i = completed_darts + 1;
        i <= options.darts;
        ++i
    ) {

        const double x =
            2.0 * rng_uniform(&rng_state) - 1.0;

        const double y =
            2.0 * rng_uniform(&rng_state) - 1.0;

        if (x * x + y * y <= 1.0) {
            ++inside_circle;
        }

        /*
         * Opportunistic signal-aware checkpoint.
         *
         * Periodic checkpoints remain the safety net for an
         * unexpected preemption with little/no warning.
         */
        if (stop_requested) {

            checkpoint_t checkpoint;

            initialize_checkpoint(&checkpoint);

            checkpoint.completed_darts = i;
            checkpoint.inside_circle = inside_circle;
            checkpoint.rng_state = rng_state;
            checkpoint.total_darts = options.darts;
            checkpoint.seed = options.seed;

            checkpoint.elapsed_seconds =
                previous_elapsed +
                wall_time() -
                start_time;

            printf(
                "\nSignal %d received. "
                "Saving checkpoint at dart %" PRIu64 "...\n",
                (int) received_signal,
                i
            );

            save_checkpoint(
                options.checkpoint_file,
                &checkpoint
            );

            printf("Checkpoint saved. Exiting.\n");

            return EXIT_SUCCESS;
        }

        if (
            options.report_every > 0 &&
            (
                i % options.report_every == 0 ||
                i == options.darts
            )
        ) {

            const double pi_estimate =
                4.0 * (double) inside_circle /
                (double) i;

            printf(
                "darts = %12" PRIu64
                " / %" PRIu64
                "   pi = %.8f"
                "   error = %.3e\n",
                i,
                options.darts,
                pi_estimate,
                fabs(pi_estimate - PI_TRUE)
            );

            fflush(stdout);

            sleep_seconds(options.sleep_seconds);
        }

        if (
            options.checkpoint_every > 0 &&
            i % options.checkpoint_every == 0
        ) {

            checkpoint_t checkpoint;

            initialize_checkpoint(&checkpoint);

            checkpoint.completed_darts = i;
            checkpoint.inside_circle = inside_circle;
            checkpoint.rng_state = rng_state;
            checkpoint.total_darts = options.darts;
            checkpoint.seed = options.seed;

            checkpoint.elapsed_seconds =
                previous_elapsed +
                wall_time() -
                start_time;

            save_checkpoint(
                options.checkpoint_file,
                &checkpoint
            );

            printf(
                "  --> periodic checkpoint saved "
                "at dart %" PRIu64 "\n",
                i
            );

            fflush(stdout);
        }
    }

    const double total_elapsed =
        previous_elapsed +
        wall_time() -
        start_time;

    const double pi_estimate =
        4.0 * (double) inside_circle /
        (double) options.darts;

    checkpoint_t checkpoint;

    initialize_checkpoint(&checkpoint);

    checkpoint.completed_darts = options.darts;
    checkpoint.inside_circle = inside_circle;
    checkpoint.rng_state = rng_state;
    checkpoint.total_darts = options.darts;
    checkpoint.seed = options.seed;
    checkpoint.elapsed_seconds = total_elapsed;

    save_checkpoint(
        options.checkpoint_file,
        &checkpoint
    );

    printf("\n");
    printf("========================================================================\n");
    printf("CALCULATION COMPLETE\n");
    printf("========================================================================\n");
    printf("Estimated pi       : %.10f\n", pi_estimate);
    printf("Reference pi       : %.10f\n", PI_TRUE);
    printf(
        "Absolute error     : %.6e\n",
        fabs(pi_estimate - PI_TRUE)
    );
    printf("Total elapsed time : %.2f s\n", total_elapsed);
    printf("========================================================================\n");

    return EXIT_SUCCESS;
}
